### PR TITLE
Canvas: Avoid conflicting stylesheets when loading SVG icons

### PR DESCRIPTION
**Pull request by adela-almasan**
While investigating https://github.com/grafana/support-escalations/issues/7316 we noticed that the SVGs used had the same class names defined between the `<style>` tag, which was causing the SVGs to borrow styling from one to another. The issue doesn't happen for SVGs that don't have custom styling defined. 
The solution was to add a unique ID to each SVG and use the ID when defining the classes.

Before

https://github.com/grafana/grafana/assets/88068998/3032dc27-7776-47ad-8a3a-6af378004bf6

After

https://github.com/grafana/grafana/assets/88068998/59079111-b0e9-4de2-b347-70dbcd9bdfd1

Fixes https://github.com/grafana/support-escalations/issues/7316

**Special notes for your reviewer:**
Cases covered:
- add new ID to SVG when not defined
- if ID is defined, use it/overwrite the existing one in `<style>`

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: public/app/features/canvas/elements/icon.tsx [hunk 1]
```diff
@@ -60,6 +60,7 @@ export function IconDisplay(props: CanvasElementProps) {
       src={data.path}
       style={svgStyle}
       className={svgStyle.strokeWidth ? svgStrokePathClass : undefined}
+      cleanStyle={true}
```
Comment (nmarrs) [inline public/app/features/canvas/elements/icon.tsx:63]: Hi just checking in from the alps 😬 we might want to also test other places we use custom svgs (like geomap) great job finding a viable fix!

File: public/app/core/components/SVG/SanitizedSVG.tsx [hunk 2]
```diff
@@ -13,7 +20,31 @@ function getCleanSVG(code: string): string {
...
+      clean = svgStyleCleanup(clean);
+    }
+
     cache.set(code, clean);
   }
+
   return clean;
 }
+
+function svgStyleCleanup(elementCode: string) {
```
[the remark this answers is not in the document]
Comment (adela-almasan) [reply public/app/core/components/SVG/SanitizedSVG.tsx:34]: yes - on it!

**Comment by adela-almasan** [comment]
tested in Geomap and it doesn't seem to be affected, the markers are keeping their colors, as their should. However, I can't use the SVGs from the escalation, not sure if there's something wrong with them - but they don't work regardless of my changes. @drew08t any idea?

https://github.com/grafana/grafana/assets/88068998/621145c3-e337-4cc6-ae85-51a5065a4b07

File: public/app/core/components/SVG/SanitizedSVG.tsx [hunk 3]
```diff
@@ -1,10 +1,19 @@
...
-export const SanitizedSVG = (props: Props) => {
-  return <SVG {...props} cacheRequests={true} preProcessor={getCleanSVG} />;
+import { getSvgId, getSvgStyle } from './utils';
+
+type SanitizedSVGProps = Props & { cleanStyle?: boolean };
+let shouldCleanSvgStyle = false;
+
+export const SanitizedSVG = (props: SanitizedSVGProps) => {
+  const { cleanStyle, ...inlineSvgProps } = props;
+  shouldCleanSvgStyle = cleanStyle ?? false;
```
Comment (ryantxu) [inline public/app/core/components/SVG/SanitizedSVG.tsx:14]: :( 

looking closer... this does not make sense -- we can not have a global state toggled by the individual component!

I think we need two `getCleanSVG` functions -- one that cleans and another that does not

File: public/app/core/components/SVG/utils.ts [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+export const getSvgId = (svgCode: string) => {
+  return svgCode.match(new RegExp('<svg.*id\\s*=\\s*([\'"])(.*?)\\1'))?.[2];
+};
+
+export const svgStyleCleanup = (svgCode: string) => {
+  let svgId = getSvgId(svgCode);
+  if (!svgId) {
+    svgId = `x${uuidv4()}`;
+    const pos = svgCode.indexOf('<svg') + 5;
```
Comment (nmarrs) [inline public/app/core/components/SVG/utils.ts:16]: any reason in particular for +5? We might want to consider replacing magic number with variable (something like `POSITION_OFFSET`?)

File: public/app/core/components/SVG/utils.ts [hunk 5]
```diff
@@ -0,0 +1,27 @@
+import { v4 as uuidv4 } from 'uuid';
+
+export const getSvgStyle = (svgCode: string) => {
+  const svgStyle = svgCode.match(new RegExp('<style type="text/css">([\\s\\S]*?)<\\/style>'));
+  return svgStyle ? svgStyle[0] : null;
+};
+
+export const getSvgId = (svgCode: string) => {
+  return svgCode.match(new RegExp('<svg.*id\\s*=\\s*([\'"])(.*?)\\1'))?.[2];
```
Comment (nmarrs) [inline public/app/core/components/SVG/utils.ts:9]: This looks a little scary to maintain 😬 at very least maybe we should consider replacing the `2` with a const explaining what the index represents?

File: public/app/core/components/SVG/SanitizedSVG.test.tsx [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+  });
+
+  it('should cleanup the style and use the existing ID', () => {
+    const cleanStyle = svgStyleCleanup(svgWithId);
+    const updatedStyle = getSvgStyle(cleanStyle);
+    const svgId = getSvgId(cleanStyle);
+
+    expect(cleanStyle.indexOf('id="')).toBeGreaterThan(-1);
+    expect(svgId).toBeDefined();
+    expect(svgId).toEqual('TEST_ID');
```
Comment (nmarrs) [inline public/app/core/components/SVG/SanitizedSVG.test.tsx:33]: `TEST_ID` should probably be a separate const as it's being used in more than 1 spot

**Review by nmarrs — APPROVED** [review]
A few nits - overall looks good and tested locally 💯

**Pull request merged**